Instantiate polymorphic qualifiers from type argument bounds

aosen-xiong opened the pull request:
Fixes eisop#1074.
Fixes eisop#1819.

This updates polymorphic qualifier instantiation for receiver type arguments whose actual type is a wildcard or type variable.

Previously, `AbstractQualifierPolymorphism` derived the instantiation from the current type node's annotation. For wildcard/type-variable type arguments, that can miss the qualifier that determines readable values. For example, in `C<? extends T>` where `T extends @Nullable Object`, `@PolyNull` should instantiate from `T`'s upper bound as `@Nullable`, not from the type-variable node itself.

The new logic follows the extends/upper-bound chain for wildcards and type variables, with a visited set to avoid recursive-bound loops, then falls back to the existing effective annotation behavior for concrete types.

aosen-xiong committed:
Add test case for crash (expecting error)

aosen-xiong committed:
Get extends bound annotation

aosen-xiong committed:
Better test coverage

aosen-xiong committed:
Merge branch 'master' into poly-typeargs

Ao-senXiong commented:
@wmdietl 
For this code,
```
class MyCollection<E> { 
  void copyOf(MyCollection<? extends E> elements) {
    Object[] array = elements.toArray();
  }
  Object [] toArray(MyCollection<@PolyVP E> this) {
      return null;
  };
}
```

Currently, `@PolyVP E` as a type argument is applying polymorphic qualifier to both bounds, i.e. `@PolyVP void` as lower bound and `@PolyVP Object` as upper bound. However, `elements` is using  `@Bottom void` as lower bound and `@Top Object` as upper bound. There is a mismatch in polymorphic qualifier substitution as in type argument we can not specify lower and upper bound seperately. We need to neither reject the code or use one bound conservatively.

aosen-xiong committed:
Merge branch 'master' into poly-typeargs

aosen-xiong committed:
Merge branch 'master' into poly-typeargs

aosen-xiong committed:
Merge branch 'master' into poly-typeargs

aosen-xiong committed:
Merge branch 'master' into poly-typeargs

aosen-xiong committed:
Handle poly instantiation from type argument bounds

aosen-xiong committed:
Merge branch 'master' into poly-typeargs

aosen-xiong committed:
Merge branch 'master' into poly-typeargs

wmdietl committed:
Merge branch 'master' into poly-typeargs

wmdietl commented on framework/src/main/java/org/checkerframework/framework/type/poly/AbstractQualifierPolymorphism.java:
```diff
@@ -347,7 +347,7 @@ private AnnotationMirrorMap<AnnotationMirror> mapQualifierToPoly(
             AnnotationMirror poly = kv.getKey();
             if (polyType.hasAnnotation(poly)) {
-                AnnotationMirror typeQual = type.getAnnotationInHierarchy(top);
+                AnnotationMirror typeQual = getInstantiationQualifier(type, top);
```
```suggestion
                AnnotationMirror typeQual = type.getEffectiveAnnotationInHierarchy(top);
```

Do all tests work with this? When do we need all the functionality in `getInstantiationQualifier`?

aosen-xiong replied:
Yes, all tests pass with getEffectiveAnnotationInHierarchy, so getInstantiationQualifier is unnecessary.

aosen-xiong committed:
Use getEffectiveAnnotationInHierarchy for poly instantiation

getErased already walks wildcard extends and type-variable upper bounds,
so the explicit getInstantiationQualifier helper was redundant.

aosen-xiong committed:
Merge branch 'master' into poly-typeargs

wmdietl reviewed (approved):
Much simpler, thanks!
